File: services/media-sets-service/src/metrics.rs
//! Prometheus metric families exported by the service.
//!
//! Pre-registered at boot so `/metrics` always surfaces the prefix even
//! before the first request lands.
//!
//! ## Foundry cost-model alignment
//!
//! The four `media_*` families below mirror the per-domain metrics
//! Foundry uses in *Media usage costs and limits* (compute seconds per
//! transformation, live-storage bytes per schema, in-flight transactions,
//! retention purges). They are the inputs operators feed into the
//! tenant cost model. Keep names and label sets stable — Grafana
//! dashboards and the Prometheus alert rules in
//! [`infra/k8s/platform/observability/prometheus-rules/media-sets.yaml`]
//! depend on them.

use once_cell::sync::Lazy;
use prometheus::{
    IntCounter, IntCounterVec, IntGauge, IntGaugeVec, register_int_counter,
    register_int_counter_vec, register_int_gauge, register_int_gauge_vec,
};

pub static MEDIA_SET_UPLOADS_TOTAL: Lazy<IntCounter> = Lazy::new(|| {
    register_int_counter!(
        "media_set_uploads_total",
        "Total number of media items registered for upload (presigned URL issued)."
    )
    .expect("register media_set_uploads_total")
});

pub static MEDIA_SET_DOWNLOADS_TOTAL: Lazy<IntCounter> = Lazy::new(|| {
    register_int_counter!(
        "media_set_downloads_total",
        "Total number of presigned download URLs issued."
    )
    .expect("register media_set_downloads_total")
});

pub static MEDIA_SET_STORAGE_BYTES: Lazy<IntGauge> = Lazy::new(|| {
    register_int_gauge!(
        "media_set_storage_bytes",
        "Approximate total bytes of media items currently live (sum of size_bytes \
         where deleted_at IS NULL)."
    )
    .expect("register media_set_storage_bytes")
});

/// Per-`(transformation, schema)` count of compute seconds consumed by
/// access-pattern work (image transform, OCR, transcription, …). Base
/// counter for the per-tenant cost model derived from Foundry's
/// "Media usage costs and limits". Incremented as access-pattern
/// handlers land — kept registered today so dashboards do not
/// disappear when the surface ships.
///
/// Labels:
/// * `transformation` — short identifier of the access pattern
///   (`image_transform`, `ocr`, `transcription`, …).
/// * `schema` — media-set schema (`IMAGE`, `AUDIO`, `VIDEO`, …).
pub static MEDIA_COMPUTE_SECONDS_TOTAL: Lazy<IntCounterVec> = Lazy::new(|| {
    register_int_counter_vec!(
        "media_compute_seconds_total",
        "Compute seconds consumed by media access-pattern work, per transformation and schema. \
         Base counter for the tenant cost model.",
        &["transformation", "schema"]
    )
    .expect("register media_compute_seconds_total")
});

/// Per-`(schema, virtual)` live storage footprint. Foundry's cost
/// model bills virtual sets at 0 because Foundry never holds the bytes
/// — surfaced separately so the dashboard can split the chargeable
/// vs reference storage.
pub static MEDIA_STORAGE_BYTES_LABELLED: Lazy<IntGaugeVec> = Lazy::new(|| {
    register_int_gauge_vec!(
        "media_storage_bytes",
        "Live media bytes by media-set schema and virtual flag. \
         Sum of size_bytes where deleted_at IS NULL.",
        &["schema", "virtual"]
    )
    .expect("register media_storage_bytes")
});

/// Open transactions per media set. Surface for the operator to spot
/// stuck writers (transactions that never commit / abort), and for the
/// alert rule to page on a sustained backlog.
pub static MEDIA_ACTIVE_TRANSACTIONS: Lazy<IntGaugeVec> = Lazy::new(|| {
    register_int_gauge_vec!(
        "media_active_transactions",
        "Currently OPEN media-set transactions, by media set RID.",
        &["media_set"]
    )
    .expect("register media_active_transactions")
});

/// Number of items soft-deleted by the retention reaper. Anomalous
/// spikes are the canonical signal for "retention policy reduced
/// accidentally" — wired to the `MediaSetsRetentionPurgeAnomaly`
/// alert in `media-sets.yaml`.
pub static MEDIA_RETENTION_PURGES_TOTAL: Lazy<IntCounter> = Lazy::new(|| {
    register_int_counter!(
        "media_retention_purges_total",
        "Total number of media items soft-deleted by the retention reaper."
    )
    .expect("register media_retention_purges_total")
});

/// Force evaluation of every metric family so `/metrics` lists them
/// before the first request that would otherwise lazily register them.
pub fn init() {
    let _ = &*MEDIA_SET_UPLOADS_TOTAL;
    let _ = &*MEDIA_SET_DOWNLOADS_TOTAL;
    let _ = &*MEDIA_SET_STORAGE_BYTES;
    let _ = &*MEDIA_COMPUTE_SECONDS_TOTAL;
    let _ = &*MEDIA_STORAGE_BYTES_LABELLED;
    let _ = &*MEDIA_ACTIVE_TRANSACTIONS;
    let _ = &*MEDIA_RETENTION_PURGES_TOTAL;
}

File: services/media-sets-service/internal/metrics/metrics.go
// Package metrics declares the Prometheus families specific to
// media-sets-service. Mirrors libs/services/media-sets-service/src/metrics.rs:
// names, labels and help text are byte-for-byte identical so dashboards
// and the Prometheus rules in
// infra/k8s/platform/observability/prometheus-rules/media-sets.yaml
// keep working unchanged.
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"

	"github.com/openfoundry/openfoundry-go/libs/observability"
)

// Metrics holds the families used by access-pattern + media-item paths.
// Created once at boot, registered on observability.Metrics.Registry,
// then threaded through to handlers/services.
type Metrics struct {
	// MediaComputeSecondsTotal is the per-(transformation, schema)
	// counter of compute-seconds consumed by access-pattern work.
	// Base counter for the tenant cost model; rates come from
	// libs/observability/costmodel.
	MediaComputeSecondsTotal *prometheus.CounterVec
	// MediaActiveTransactions is the per-media-set gauge of currently
	// OPEN transactions. The transactions service bumps it on open
	// and decrements it on close (commit/abort).
	MediaActiveTransactions *prometheus.GaugeVec
	// MediaRetentionPurgesTotal is the global counter of items the
	// retention reaper soft-deletes. Anomalous spikes are the
	// canonical "retention reduced accidentally" signal.
	MediaRetentionPurgesTotal prometheus.Counter
}

// New registers the families on the given observability.Metrics
// registry and returns the typed handle. Pre-evaluates each family so
// /metrics surfaces the prefix even before the first invocation.
func New(o *observability.Metrics) *Metrics {
	m := &Metrics{
		MediaComputeSecondsTotal: prometheus.NewCounterVec(
			prometheus.CounterOpts{
				Name: "media_compute_seconds_total",
				Help: "Compute seconds consumed by media access-pattern work, per transformation and schema. Base counter for the tenant cost model.",
			},
			[]string{"transformation", "schema"},
		),
		MediaActiveTransactions: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Name: "media_active_transactions",
				Help: "Currently OPEN media-set transactions, by media set RID.",
			},
			[]string{"media_set"},
		),
		MediaRetentionPurgesTotal: prometheus.NewCounter(
			prometheus.CounterOpts{
				Name: "media_retention_purges_total",
				Help: "Total number of media items soft-deleted by the retention reaper.",
			},
		),
	}
	o.Register(m.MediaComputeSecondsTotal)
	o.Register(m.MediaActiveTransactions)
	o.Register(m.MediaRetentionPurgesTotal)
	return m
}
